### ISSUE TITLE
.at set unexpected behaviour

### ISSUE THREAD

**Issue by afshawnlotfi** [google/jax#19373]
### Description

Correct Behaviour:
```python
test_arr = jnp.arange(7*2*3, dtype=jnp.float32).reshape(7,2,3)
jnp.zeros((7,2,3)).at[np.array([0,1,6,3,4,5,2], dtype=np.int32)].set(test_arr)
```

Incorrect Behaviour:
```python
test_arr = jnp.arange(7*2*3, dtype=jnp.float32).reshape(7,2,3)
jnp.zeros((7,2,3)).at[np.array([0,2,6,3,4,5,1], dtype=np.int32)].set(test_arr)
```

Why isn't the highlighted 
<img width="360" alt="image" src="https://github.com/google/jax/assets/6283745/59c01a80-113e-4e3c-aec0-d271fe3cb90b">

This:
```
 [[12. 13. 14.]
  [15. 16. 17.]]
```

And onwards ...

### What jax/jaxlib version are you using?

0.4.20 0.4.20

### Which accelerator(s) are you using?

CPU

### Additional system info?

MacOS 14.1 (23B74)

### NVIDIA GPU info

N/A

**Comment by jakevdp** [comment]
Thanks for the report – I'm able to reproduce the issue. I'll take a look today.

**Comment by jakevdp** [comment]
Confirmed that this error arises on CPU and GPU (no it's likely not some compiler optimization gone awry) and is present back as far as jax v0.4.1 (not something that's recently regressed)

**Comment by jakevdp** [comment]
Wait, it looks like this is actually working as expected: for example this:
```python
jnp.zeros((7,2,3)).at[np.array([0,2,6,3,4,5,1], dtype=np.int32)].set(test_arr)
```
is similar to writing this in numpy:
```
x = np.zeros((7, 2, 3))
x[np.array([0,2,6,3,4,5,1], dtype=np.int32)] = np.array(test_arr)
print(x)
```
```

[[[ 0.  1.  2.]
  [ 3.  4.  5.]]

 [[36. 37. 38.]
  [39. 40. 41.]]

 [[ 6.  7.  8.]
  [ 9. 10. 11.]]

 [[18. 19. 20.]
  [21. 22. 23.]]

 [[24. 25. 26.]
  [27. 28. 29.]]

 [[30. 31. 32.]
  [33. 34. 35.]]

 [[12. 13. 14.]
  [15. 16. 17.]]]
```
The indices are broadcast with the leading dimension of `test_arr`, so index `1` corresponds to the last entry in the leading dimension of the array, so we expect the block starting with `36` to be in entry `1` of the result.

Does that make sense?

**Comment by afshawnlotfi** [comment]
I'm having trouble understanding the broadcast over the leading edge part. Ideally I'd want to sort the indexes by axis=0 here. 

And the reason I don't just mask set is I actually have two arrays with indexes that I want to properly sort into one array for context.

**Comment by afshawnlotfi** [comment]
Oh I see it's doing it in reverse

**Comment by jakevdp** [comment]
Maybe it's simpler to see it this way. Here's a 1D version:
```python
x = jnp.arange(7)
idx = np.array([0,2,6,3,4,5,1], dtype=np.int32)
print(jnp.zeros(7).at[idx].set(x))
# [0. 6. 1. 3. 4. 5. 2.]
```
This is effectively a batched version of this:
```python
index 0 set to 0
index 2 set to 1
index 6 set to 2
index 3 set to 3
index 4 set to 4
index 5 set to 5
index 1 set to 6
[0. 6. 1. 3. 4. 5. 2.]
```
From your question, it sounds like you were expecgint index 1 to be set to 2, but index 1 is being (correctly) set to 6. Does that make sense?

**Comment by jakevdp** [comment]
Perhaps instead of
```python
result = jnp.zeros((7,2,3)).at[np.array([0,1,6,3,4,5,2], dtype=np.int32)].set(test_arr)
```
you were hoping to compute this?
```python
result = test_arr[np.array([0,1,6,3,4,5,2], dtype=np.int32)]
```

**Comment by afshawnlotfi** [comment]
Appreciate it!
